### PR TITLE
Add Rope Implementation

### PR DESCRIPTION
**Pull request by jainapurva**
## Summary

Adds Rotary Position Embedding (RoPE) as a functional API in `torch.nn.functional`. The implementation uses standard PyTorch ops that are efficiently fused by `torch.compile` with Inductor — no custom CUDA kernels needed.

### API

- **`F.rotary_embedding_frequencies(dim, end, theta=10000.0)`** — Precompute cos/sin frequencies for RoPE (NeoX-style layout)
- **`F.apply_rotary_emb(query, key, cos, sin, seq_dim=1, interleaved=False)`** — Apply rotary embedding to query/key tensors. Supports BSHD (`seq_dim=1`) and BHSD (`seq_dim=2`) layouts, plus NeoX and GPT-J interleaving styles
- **`F.rotate_half(x, interleaved=False)`** — Rotate half dimensions (building block for RoPE)

For custom frequency scaling (YaRN, LLaMA3, NTK, etc.), users compute their own cos/sin and pass them directly to `apply_rotary_emb`.

### Profiler Results — NVIDIA B200

**Config:** `batch=4, seq=2048, heads=32, dim=128, dtype=bfloat16`

| Mode | CUDA Time (10 iters) | Kernel Launches | Fused? |
|------|---------------------|-----------------|--------|
| Eager (no compile) | **8.675 ms** | 100 (mul, neg, cat, add) | No — separate kernels |
| `torch.compile` | **0.470 ms** | 10 | Yes — single `triton_poi_fused_add_cat_mul_neg_split_unsqueeze_0` |
| **Speedup** | **~18.5x** | **10x fewer launches** | |

`torch.compile` fuses ALL RoPE operations (mul, neg, cat, add, chunk, unsqueeze, split) into a **single Triton kernel**.

<details>
<summary>Eager mode profiler output (click to expand)</summary>

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                     F.apply_rotary_emb         0.00%       0.000us         0.00%       0.000us       0.000us       8.869ms       102.23%       8.869ms     886.875us            10
                                     F.apply_rotary_emb         8.11%     814.081us        32.22%       3.235ms     323.507us       0.000us         0.00%       8.773ms     877.312us            10
                                              aten::mul         2.59%     259.848us        15.05%       1.511ms      37.781us       3.815ms        43.98%       3.913ms      97.823us            40
                                              aten::cat         1.30%     130.935us         2.02%     203.075us      10.154us       3.482ms        40.14%       3.482ms     174.120us            20
                                              aten::neg         1.01%     101.117us         1.75%     176.102us       8.805us     728.830us         8.40%     728.830us      36.442us            20
                                              aten::add         0.86%      86.378us         1.49%     149.165us       7.458us     648.960us         7.48%     648.960us      32.448us            20
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 10.041ms
Self CUDA time total: 8.675ms
```

</details>

<details>
<summary>torch.compile profiler output (click to expand)</summary>

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                     F.apply_rotary_emb        14.05%     390.440us        99.43%       2.763ms     276.316us       0.000us         0.00%     470.117us      47.012us            10
                             Torch-Compiled Region: 0/0        12.68%     352.249us        82.92%       2.304ms     230.424us       0.000us         0.00%     470.117us      47.012us            10
     triton_poi_fused_add_cat_mul_neg_split_unsqueeze_0         4.39%     122.053us         7.63%     212.068us      21.207us     470.117us       100.00%     470.117us      47.012us            10
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 2.779ms
Self CUDA time total: 470.117us
```

</details>

### Perfetto Trace Links

- [Eager mode trace (F.apply_rotary_emb — separate kernels)](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/appy_844ca0f7-b76e-4e20-bd4a-7c432c4622c2_rope_trace_eager.json)
- [torch.compile trace (F.apply_rotary_emb — fused Triton kernel)](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/appy_d1dec731-9fbc-486a-8ba9-cafe251f87a3_rope_trace_compile.json)

### Validated on Real Models

Replaced custom complex-number RoPE in SAM3 and SAM2 with this API. Both models
use 2D axial RoPE (interleaved/complex layout), which maps to
`F.apply_rotary_emb(..., interleaved=True)`.

**SAM3** — [jainapurva/sam3#1](https://github.com/jainapurva/sam3/pull/1)

E2E inference on NVIDIA B200:

| Metric | Baseline (complex arithmetic) | `F.apply_rotary_emb` |
|---|---|---|
| `complex MulFunctor` kernel calls | 64 (0.631ms) | **0** |
| `view_as_complex` / `view_as_real` | 128 calls | **0** |
| `copy_` calls | 850 (4.459ms) | 786 |
| Peak memory delta | 316.6 MB | 316.5 MB |
| Inference output | 12 masks, scores 0.87–0.90 | **identical** |

All 192 complex-number kernel calls eliminated. No memory overhead. Numerically
identical outputs. `use_rope_real` workaround for `torch.compile` is no longer
needed.

**SAM2** — [pytorch/ao#4118](https://github.com/pytorch/ao/pull/4118)

Numerical equivalence tests (10/10 pass, max diff 4.77e-07 in fp32):
basic attention, cross-attention with `rope_k_repeat`, `num_k_exclude_rope`,
`torch.compile(fullgraph=True)`, and bfloat16.

### Files Changed

- `torch/nn/functional.py` — RoPE implementation (3 functions)
- `torch/overrides.py` — Register functions in ignored overrides
- `docs/source/nn.functional.rst` — Documentation entries
- `test/test_transformers.py` — TestRoPE class (shape, numerical correctness, gradcheck, norm preservation)
- `test/test_fx.py` — FX tracing support + pickle test

## Test plan

- [x] `TestRoPE.test_rope_shape` — Verifies output shapes for BSHD/BHSD layouts + interleaved modes
- [x] `TestRoPE.test_rope_frequencies_odd_dim_error` — Validates error on odd dimensions
- [x] `TestRoPE.test_rope_numerical_correctness` — Compares against HuggingFace reference for float32/float16/bfloat16
- [x] `TestRoPE.test_rope_gradcheck` — Gradient correctness via `torch.autograd.gradcheck`
- [x] `TestRoPE.test_rope_norm_preservation` — Verifies RoPE preserves vector norms (rotation property)
- [x] FX tracing: `apply_rotary_emb` and `rotate_half` registered in `TestFunctionalTracing`
- [x] Profiled on NVIDIA B200: ~18.5x CUDA speedup with torch.compile kernel fusion

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -5099,6 +5099,87 @@ def test_is_causal_and_mask_fails(self, device):
         with self.assertRaisesRegex(ValueError, "CausalBias should not be used with causal=True"):
             scaled_dot_product_attention(query, key, value, attn_mask=attn_bias, is_causal=True, dropout_p=0.0)
 
+class TestRoPE(NNTestCase):
```
Comment (jbschlosser) [inline test/test_transformers.py:5225]: I think it'd be nice to have some tests with compile, especially ensuring RoPE application can be epilogue-fused with transformer input projection.

Also more generally, I think it's important to have a set of tests that are representative of how we expect RoPE to be used within a transformer context. In particular, I'm interested in ensuring freqs can be precomputed and ragged-sliced (for varying seqlens across the batch) in a compile-compatible way.

**Review by jbschlosser — COMMENTED** [review]
Thanks for the hard work here! The individual pieces look generally correct to me, but in my head there's a gap on ensuring these fit within appropriately within a transformer context. They may very well work fine with compile, but I'd like to see more thorough testing ensuring this. Thanks again for looking into it!
